modules/apps/batch-engine/batch-engine-service/src/main/java/com/liferay/batch/engine/internal/writer/ColumnValueWriter.java

```diff
@@ -0,0 +1,103 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.batch.engine.internal.writer;
+
+import com.liferay.petra.string.StringPool;
+
+import java.lang.reflect.Field;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.function.Consumer;
+
+/**
+ * @author Ivica cardic
+ */
+public class ColumnValueWriter {
+
+	public void write(
+			Consumer<Collection<?>> consumer, Map<String, Field> fieldMap,
+			List<String> fieldNames, Object item)
+		throws IllegalAccessException {
+
+		Map<String, Object> columnNameValueMap = _getColumnNameValueMap(
+			item, fieldMap, fieldNames);
+
+		if (!_firstLineWritten) {
+			consumer.accept(fieldNames);
```

Comment:
@ivicac there is a major issue here.
Headers are written in the order of given fieldNames list, but the following values are written in the order of sorted keys, which could be totally different.

Reply:
And please update the unit test to verify this. Right now in the test you are feeding the fieldNames sorted, which just happen to make the output right. If shuffling the order, the output will be wrong.

Reply:
Not only that, if there is an invalid fieldName, it will be added to headers, but no value will never be outputted for it, then the output file will still have a broken format.

Basically when doing the value extraction, the loop must be done based on "fieldNames", not based on "fieldMap".